### PR TITLE
subsys: bootloader: fprotect: Add BPROT implementation

### PR DESCRIPTION
**Pull request by oyvindronningstad**
Implement fprotect_bprot.c (use NRF_BPROT directly, no HAL yet).
Also add renaming to make the BPROT implementation work for nRF51.

Together with #333 this makes nRF52832 work, but nRF51 and nRF52810 don't compile yet.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

### CODE REVIEW

**Comment by oyvindronningstad** [comment]
cc @hakonfam @sigvartmh @lemrey @nvlsianpu @carlescufi

File: subsys/bootloader/fprotect/CMakeLists.txt [hunk 1]
```diff
@@ -10,7 +10,7 @@ if (CONFIG_SB_FLASH_PROTECT)
     set(FPROTECT_SRC fprotect_spu.c)
   elseif (CONFIG_HAS_HW_NRF_ACL)
     set(FPROTECT_SRC fprotect_acl.c)
-  elseif (CONFIG_HAS_HW_BPROT)
+  elseif (CONFIG_HAS_HW_NRF_BPROT OR CONFIG_HAS_HW_NRF_MPU)
```
Comment (hakonfam) [inline subsys/bootloader/fprotect/CMakeLists.txt]: Do we use bprot if MPU is in place?

Comment (sigvartmh) [reply subsys/bootloader/fprotect/CMakeLists.txt]: MPU and BPROT has the same interface

Comment (oyvindronningstad) [reply subsys/bootloader/fprotect/CMakeLists.txt]: I got the idea from nrf51_to_nrf52.h

File: subsys/bootloader/fprotect/fprotect_bprot.c [hunk 2]
```diff
@@ -4,7 +4,62 @@
...
+	u32_t pagenum_end   = (start + length) / BPROT_REGIONS_SIZE;
+
+	if ((start % BPROT_REGIONS_SIZE) ||
+	    (length % BPROT_REGIONS_SIZE) ||
+	    (pagenum_end >= BPROT_REGIONS_NUM)) {
+		/* start or length isn't aligned with a BPROT region,
+		 * or attempting to protect an area outside flash. */
+		return -EFAULT;
+	}
+
```
Comment (pdunaj) [inline subsys/bootloader/fprotect/fprotect_bprot.c]: `if (pagenum_end <= pagenum_start) return -EINVAL`?

**Comment by hakonfam** [comment]
I have updated https://github.com/NordicPlayground/fw-nrfconnect-nrf/pull/336 to use `-EINVAL`  instead of `-EFAULT` when there is an invalid argument, you should use the same here :)

File: subsys/bootloader/fprotect/fprotect_bprot.c (hunk 2 above)
Comment (oyvindronningstad) [reply subsys/bootloader/fprotect/fprotect_bprot.c]: Good catch :). Changed retval to -EINVAL and added the check.

**Comment by oyvindronningstad** [comment]
> I have updated #336 to use `-EINVAL` instead of `-EFAULT` when there is an invalid argument, you should use the same here :)

Done

**Comment by nvlsianpu** [comment]
It is the same as #303 was.

**Comment by oyvindronningstad** [comment]
> It is the same as #303 was.

Yes. #303 was messed up by the release and the forks going public.

**Comment by oyvindronningstad** [comment]
@carlescufi can we merge this?

**Pull request merged**